src/crimson/os/seastore/object_data_handler.cc

```diff
@@ -551,6 +551,78 @@ ObjectDataHandler::read_ret ObjectDataHandler::read(
     });
 }
 
+ObjectDataHandler::fiemap_ret ObjectDataHandler::fiemap(
+  context_t ctx,
+  objaddr_t obj_offset,
+  extent_len_t len)
+{
+  return seastar::do_with(
+    std::map<uint64_t, uint64_t>(),
+    [ctx, obj_offset, len](auto &ret) {
+    return with_object_data(
+      ctx,
+      [ctx, obj_offset, len, &ret](const auto &object_data) {
+      LOG_PREFIX(ObjectDataHandler::sparse_read);
+      DEBUGT("sparse_reading {}~{}",
+              ctx.t,
+              object_data.get_reserved_data_base(),
+              object_data.get_reserved_data_len());
+      /* Assumption: callers ensure that onode size is <= reserved
+       * size and that len is adjusted here prior to call */
+      ceph_assert(!object_data.is_null());
+      ceph_assert((obj_offset + len) <= object_data.get_reserved_data_len());
+      ceph_assert(len > 0);
+      laddr_t loffset =
+        object_data.get_reserved_data_base() + obj_offset;
+      return ctx.tm.get_pins(
+        ctx.t,
+        loffset,
+        len
+      ).si_then([ctx, loffset, len, &ret, &object_data](auto _pins) {
+        ceph_assert(_pins.size() >= 1);
+        ceph_assert((*_pins.begin())->get_laddr() <= loffset);
+        return seastar::do_with(
+          std::move(_pins),
+          loffset,
+          len,
+          [ctx, loffset, len, &ret, &object_data]
+          (auto &pins, auto &current_loff, auto &current_len) {
+          return trans_intr::do_for_each(
+            pins,
+            [loffset, len, &current_loff, &current_len, &ret,
```

Comment:
I don't think we need to be modifying current_loff and current_len, it should be enough to simply adjust the left or right bounds of each pin as we add it to ret.